### PR TITLE
Feat : Export Specific Bookmarks

### LINKED ISSUE #90: Add Option to Export Specific Bookmark, like exporting all bookmark
**Is your feature request related to a problem? Please describe.**
Many time users doesn't want to export all bookmark, allow them to export the specific bookmark them wants.

**Describe the solution you'd like**
Allow specific bookmarks to be exported.

This issue can be broken down into two tasks:
- [x] Implement Export Specific Functionality
- [x] Enhance UI of Export Specific 

**Additional context**
PS: I want to start working on Implementing Export Specific Functionality.

### PR DESCRIPTION
**Pull request by AmanR28**
Fixes #90 

**Description**
Implements feature to export specific bookmarks to get exported.

**Other information**

**Checklist:**
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `master` branch of the repository.
- [x] My commit messages follow [best practices][best_practices].
- [ ] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.

### CODE REVIEW

**Review by makkoncept — CHANGES_REQUESTED** [review]
@AmanR28 I have tested the PR locally. It is working good, but we can do much better as far as the UX is concerned. Now the user has to go to the options page to mark the bookmarks and export them. Instead, we can do all that on the bookmarks section of the popup itself.

- On hovering the thumbnails, along with other icons, a select checkbox will also be there.
- Two more buttons will be there(along with "Delete all" button).
	- "Export all"
	- "Export selected"
- All the appropriate error handling and notification. eg: "No bookmarks selected" when "Export selected" button is clicked without selecting any bookmark images.
- Make sure that the UI is minimal yet beautiful.
- Modify the bookmarks tab in the options page to only have import bookmarks option.

**Comment by AmanR28** [comment]
@makkoncept 
Yep, that will be better than this.

Sir, can a new issue be created for this? I will open a new PR implementing this better feature ASAP.

PS: Please merge this PR.

**Comment by makkoncept** [comment]
I think that since the context of the change is very similar to this PR, you should make the changes here only. Also, the user base is growing rapidly and some important bugs are being reported. Therefore, the master should be in the condition to push a new update without any bottleneck.
Also, please don't call me "sir".
Thanks!

**Comment by AmanR28** [comment]
@makkoncept Will u please help me why i am getting that "elements.selectAllButton" is undefined, but when i am running the code it works fine and it is defined.

**Comment by AmanR28** [comment]
- [x] On hovering the thumbnails, along with other icons, a select checkbox will also be there.
- [x] Two more buttons will be there(along with "Delete all" button).
      "Export all"
      "Export selected"
- [x] All the appropriate error handling and notification. eg: "No bookmarks selected" when "Export 
       selected" button is clicked without selecting any bookmark images.
- [x] Make sure that the UI is minimal yet beautiful.

**Review by makkoncept — CHANGES_REQUESTED** [review]
@AmanR28 thanks for updating the PR. This requires a thorough review but two things that can be improved upon right away are:
- the "Select all" should have a toggle effect (clicking it would select all and then clicking it again should deselect all bookmarks). It would be good if you change the inner text of the button during the process to make it more explicit.
- The checkbox would be better if it is in the right instead of left. It would sort of balance things out.

**Comment by AmanR28** [comment]
- [x] The "Select all" should have a toggle effect (clicking it would select all and then clicking it again should deselect all bookmarks). It would be good if you change the inner text of the button during the process to make it more explicit.

- [x] The checkbox would be better if it is in the right instead of left. It would sort of balance things out.

File: src/popup/bookmarkModule.js [hunk 1]
```diff
@@ -207,6 +225,38 @@ function loadImages() {
...
+              selectedBookmarks -= 1;
+            } else {
+              btn.parentElement.setAttribute('style', 'opacity : 1');
+              selectedBookmarks += 1;
+            }
+            btn.isChecked = !btn.isChecked;
+
+            console.log(selectedBookmarks);
+            if (selectedBookmarks === elements.selectButtons.length) {
+              elements.selectAllButton.innerHTML = `
```
Comment (makkoncept) [inline src/popup/bookmarkModule.js]: Please do not use `innerHTML`([reason](https://developer.mozilla.org/en-US/docs/Web/API/Element/innerHTML#Security_considerations)). Instead use `textContent` to change the text directly inside the `.content` div.

File: src/popup/sass/components/_bookmarks.scss [hunk 2]
```diff
@@ -9,9 +9,36 @@
...
+  z-index: 3;
+  white-space: nowrap;
+  overflow: hidden;
+  text-overflow: ellipsis;
+  vertical-align: middle; 
+}
+
+.select-btn {
+  float: right;
+}
```
Comment (makkoncept) [inline src/popup/sass/components/_bookmarks.scss]: add a newline at end of file

**Review by makkoncept — CHANGES_REQUESTED** [review]
@AmanR28 I have added some review comments. Please work on them.
Also, I think we can move the checkbox so that it is horizontally aligned with the bookmark icon.
Screenshot (242)

**Comment by AmanR28** [comment]
> @AmanR28 I have added some review comments. Please work on them.

Ohk i will asap.

> Also, I think we can move the checkbox so that it is horizontally aligned with the bookmark icon.

So, should I place the checkbox to top right corner.

File: src/popup/bookmarkModule.js (hunk 1 above)
Comment (AmanR28) [reply src/popup/bookmarkModule.js]: Thanks for clarifying link.

**Comment by AmanR28** [comment]
@makkoncept  Is this alright:

Screenshot from 2020-01-25 11-07-17

File: src/popup/bookmarkModule.js [hunk 3]
```diff
@@ -207,6 +226,32 @@ function loadImages() {
...
+        .then(() => {
+          // Add onClick event to all the checkboxes
+
+          // Get checkbox data from DOM
+          const checkbox = elements.selectCheckboxes[elements.selectCheckboxes.length - 1];
+
+          checkbox.isChecked = false;
+          bookmarkDOM[checkbox.getAttribute('id')] = checkbox;
+
+          checkbox.addEventListener('click', () => {
```
Comment (makkoncept) [inline src/popup/bookmarkModule.js]: Explain what this eventlistener does

File: src/popup/bookmarkModule.js [hunk 4]
```diff
@@ -207,6 +226,32 @@ function loadImages() {
...
           appendToGrid(msnry, fragment, gridItemDiv, elements.gridBookmarks);
+        })
+        .then(() => {
+          // Add onClick event to all the checkboxes
+
+          // Get checkbox data from DOM
+          const checkbox = elements.selectCheckboxes[elements.selectCheckboxes.length - 1];
+
+          checkbox.isChecked = false;
+          bookmarkDOM[checkbox.getAttribute('id')] = checkbox;
```
Comment (makkoncept) [inline src/popup/bookmarkModule.js]: add a explanatory comment here

File: src/popup/base.js [hunk 5]
```diff
@@ -40,8 +40,9 @@ export const elements = {
   buttonBackToTop: document.getElementsByClassName('button-backToTop')[0],
   exportBookmark: document.getElementsByClassName('export-bookmarks-button')[0],
   getBookmarks: document.getElementsByClassName('grid-item'),
-  selectButtons: document.getElementsByClassName('select-btn'),
-  selectAllButton: document.getElementsByClassName('select-all-button')[0],
+  selectCheckboxes: document.getElementsByClassName('select-checkbox'),
+  selectAllCheckbox: document.getElementsByClassName('select-all-checkbox')[0],
```
Comment (makkoncept) [inline src/popup/base.js]: Make it `buttonSelectAllCheckbox` and likewise, the class should be `select-all-checkbox-button`.

File: src/popup/base.js [hunk 6]
```diff
@@ -40,8 +40,9 @@ export const elements = {
...
+  selectAllCheckboxContent: document.getElementsByClassName('select-all-checkbox-content')[0],
```
Comment (makkoncept) [inline src/popup/base.js]: There is not need to add a class `select-all-checkbox-content` to select the div. Insead use the already defined `SelectAllCheckboxButton` to select the content div.

**Review by makkoncept — CHANGES_REQUESTED** [review]
- The placement of checkbox is better than previous arrangement. You can swap the positions of the - checkbox and bookmark icon because now the placement of the checkbox icon on the images in the search section is looking off. 
Screenshot (251)
- Also, because of the new changes, I am not able to individually select or deselect the checkboxes. Only the select-all button is able to select them. Please rectify this.

**Comment by AmanR28** [comment]
@makkoncept Should I place the checkbox on the left side of bookmark icon or below it?

**Comment by AmanR28** [comment]
@makkoncept Should I just correct bookmark option on main search page, keeping the alignment of checkbox on bookmark page as I implement, i.e. on the top right most corner.
For this, I will use inline style.

**Comment by makkoncept** [comment]
@AmanR28 you can place the checkbox on the left side of the bookmark icon (essentially swapping their current places).

File: src/popup/popup.html [hunk 7]
```diff
@@ -101,6 +101,16 @@ <h3 class="text-align-center vocab heading">Bookmarks</h3>
             Delete All
           </div>
         </button>
+        <button class="select-all-checkbox-button vocab button negative-indicating dark-shaded normal-sized basic">
+          <div class="content select-all-checkbox-button-content">
```
Comment (makkoncept) [inline src/popup/popup.html]: remove the redundant class `select-all-checkbox-button-content`.

**Review by makkoncept — CHANGES_REQUESTED** [review]
@AmanR28 Looks good. Please resolve the 2 minor issues and then we can merge this.

**Review by makkoncept — APPROVED** [review]

**Pull request merged**